src/rules/no-extra-semicolons/index.js

```diff
@@ -0,0 +1,81 @@
+import {
+  ruleMessages,
+  styleSearch,
+  report,
+  validateOptions,
+} from "../../utils"
+
+export const ruleName = "no-extra-semicolons"
+
+export const messages = ruleMessages(ruleName, {
+  rejected: "Unexpected extra semicolon",
+})
+
+function getOffsetByNode(node) {
+  const string = node.root().toString()
+  const nodeColumn = node.source.start.column
+  const nodeLine = node.source.start.line
+  let line = 1
+  let column = 1
+  let index = 0
+
+  for (let i = 0; i < string.length; i++) {
+    if (column === nodeColumn && nodeLine === line) {
+      index = i
+      break
+    }
+
+    if (string[i] === "\n") {
+      column = 1
+      line  += 1
+    } else {
+      column += 1
+    }
+  }
+
+  return index
+}
+
+export default function (actual) {
+  return (root, result) => {
+    const validOptions = validateOptions(result, ruleName, { actual })
+    if (!validOptions) { return }
+
+    if (root.raws.after) {
+      const rawBeforeRoot = root.raws.after
+      styleSearch({ source: rawBeforeRoot, target: ";" }, match => {
+        complain(root.toString().length - rawBeforeRoot.length + match.startIndex)
+      })
+    }
+
+    root.walk((node) => {
+      if (node.raws.before) {
+        const rawBeforeNode = node.raws.before
+        styleSearch({ source: rawBeforeNode, target: ";" }, match => {
+          complain(getOffsetByNode(node) - rawBeforeNode.length + match.startIndex)
```

Comment:
> postcss can't get negative index by function positionBy

Ah, that's what I thought. What you've done could be a stopgap measure — I think it only runs when you have an error, anyway, so won't slow things down significantly. But it might be worth filing an issue with PostCSS so someone looks into what it would take for `positionBy` to understand negative numbers.